### PR TITLE
Clean up

### PR DESCRIPTION
**Pull request by Sentient07**
The cleanup of PR  #4570. Will rebase with the master once  #4570 gets merged. 
Tasks :
- [x] Correctly handle HostFromGpu nodes 
- [ ]  Remove the special case for Alloc / AllocEmpty from the new optimizer, and register these lifters in an "out2in" topooptimizer
- [x] Replace op_lifter with out2in
- [x] Factor out caching of Op instances 
- [ ] Check if all the optimizers used by op_lifter is being used by op_lifter_topo
- [ ] Implement backward pass.
- [ ] Remove op_lifter from EquilibriumOptimizer

@lamblin Am i missing anything in the to-do list?

### CODE REVIEW

File: theano/gpuarray/extra_ops.py [hunk 1]
```diff
@@ -468,4 +469,4 @@ def use_gpu_cumsumop(op, ctx_name, inputs, outputs):
         if axis is None:
             axis = 0
         assert isinstance(x.type, GpuArrayType)
-        return GpuCumsum(axis)(x)
+        return GpuCumsum(axis, ctx_name)(x)
```
Comment (Sentient07) [inline theano/gpuarray/extra_ops.py]: I felt it is better to pass the context_name. Will change if necessary

Comment (lamblin) [inline theano/gpuarray/extra_ops.py]: As documented [here](http://deeplearning.net/software/theano_versions/dev/extending/extending_theano_gpu.html), the best practice is to infer the context during `make_node`, not to force the context when creating the Op.
Please remove the `context_name` attribute.

**Comment by Sentient07** [comment]
In line:467 of `theano/gpuarray/type.py`, if i make a call to transfer instead of `host_from_gpu`, it gives , 
`RuntimeError: maximum recursion depth exceeded`

File: theano/gpuarray/extra_ops.py (hunk 1 above)
Comment (Sentient07) [inline theano/gpuarray/extra_ops.py]: Yeah, i'll remove that commit.

**Comment by Sentient07** [comment]
In this PR, i've removed `ShapeOptimizer` from `fast_compile`. There are some optimizations that need `ShapeOptimizer` and is a part of `fast_compile`. I'm removing those optimizations from `fast_compile`. Yet to open that PR of removing `ShapeOptimizer`. @lamblin I believe that PR should get merged before this ?

File: theano/gpuarray/opt.py [hunk 2]
```diff
@@ -55,19 +60,43 @@
...
+gpu_seqopt.register('gpuarray_graph_optimization', GraphToGPUDB(), -0.5,
+                    'fast_compile', 'fast_run', 'gpuarray')
+
 gpu_seqopt.register('gpuarray_local_optimiziations', gpu_optimizer, 1,
                     'fast_compile', 'fast_run', 'gpuarray')
 gpu_seqopt.register('gpuarray_cut_transfers', gpu_cut_copies, 2,
                     'fast_compile', 'fast_run', 'gpuarray')
 
+gpu_seqopt.register('op_lifter_topo', TopoOptimizer(gpu_topo.query('+fast_compile'), order='out_to_in'),
+                    10, 'fast_run', 'gpuarray', 'fast_compile')
```
Comment (lamblin) [inline theano/gpuarray/opt.py]: Use something between -0.5 and 1 instead of 10 here.

**Comment by lamblin** [comment]
It seems to make sense for a first draft of the backward pass (using TopoOptimizer and LocalGroupDB).
When you change the priorities so it is executed before the EquilibriumOptimizer, we can see in the profiling if there is something to gain, and if we need to optimize the implementation further.
For instance, maybe we would need to implement tracks in LocalGroupDB, like in EquilibriumDB.

**Comment by Sentient07** [comment]
> For instance, maybe we would need to implement tracks in LocalGroupDB, like in EquilibriumDB.

Could you please elaborate on this ?

**Comment by lamblin** [comment]
EquilibriumDB has "tracks", or ops to match when trying to apply its local optimizers. For instance, some local optimizers will only be tried on nodes corresponding to some Ops. It uses the Ops passed to the `local_optimizer` decorator.
For instance, if a local opt registered with `@local_optimizer([GpuFromHost, GpuToGpu, HostFromGpu])`, then the EquilibriumOptimizer will only try to apply it on nodes where `node.owner.op` is an instance of those types.
I'm not sure if the LocalGroup uses this information. If it does not, then we may need to implement that logic, like in EquilibriumOptimizer.

**Comment by Sentient07** [comment]
From what I see,LocalGroup does not use that info. LocalOptimizer uses track but I don't think LocalGroup does. I'll confirm again  once.

**Comment by Sentient07** [comment]
@nouiz The new optimizer `op_lifter_topo` doesn't show all the local optimizers as `gpu_optimizer`. Am I doing the query wrong ?

**Comment by Sentient07** [comment]
We're registering `theano.tensor.opt.local_remove_all_assert` into `gpu_optimizer` separately. Don't we need to do the same for `gpu_optimizer2` ?

File: theano/gpuarray/extra_ops.py [hunk 3]
```diff
@@ -22,8 +22,9 @@ class GpuCumsum(GpuKernelBase, Op):
     SUPPORTED_NDIMS = 3
     __props__ = ('axis',)
 
-    def __init__(self, axis):
```
Comment (nouiz) [inline theano/gpuarray/extra_ops.py]: Why this change? Just remove this commit, or explain why it is needed.

Comment (Sentient07) [inline theano/gpuarray/extra_ops.py]: oh! This should have gotten removed from the rebase. It existed before, but i have changed this behaviour in #4570 itself

Comment (Sentient07) [inline theano/gpuarray/extra_ops.py]: Yeah, this diff doesn't exist anymore. It has been removed. Earlier, i thought it was better to pass context_name

File: theano/gpuarray/opt.py [hunk 4]
```diff
@@ -56,18 +61,43 @@
...
 
+gpu_seqopt.register('gpuarray_graph_optimization', GraphToGPUDB(), -0.5,
+                    'fast_compile', 'fast_run', 'gpuarray')
+
 gpu_seqopt.register('gpuarray_local_optimiziations', gpu_optimizer, 1,
                     'fast_compile', 'fast_run', 'gpuarray')
 gpu_seqopt.register('gpuarray_cut_transfers', gpu_cut_copies, 2,
                     'fast_compile', 'fast_run', 'gpuarray')
 
+gpu_seqopt.register('op_lifter_topo', TopoOptimizer(gpu_topo.query('+fast_compile', '+fast_run', '+cudnn',
```
Comment (Sentient07) [inline theano/gpuarray/opt.py]: @nouiz By registering it here, i won't have all the local optimizers in the `gpu_topo` db registered into `gpu_seqopt`?

Comment (Sentient07) [inline theano/gpuarray/opt.py]: Also, is this is the correct way to retrieve all the local optimizers in that DB

File: theano/gpuarray/basic_ops.py [hunk 5]
```diff
@@ -647,14 +647,6 @@ def c_code_cache_version(self):
         return (9,)
 
 
-# Caching GPUAlloc
```
Comment (nouiz) [inline theano/gpuarray/basic_ops.py]: You removed the caching, this is good, but to be useful, you should implement the method Op.\__new__() that do the caching automatically based on \__props__. Then we should check all subclass of Op to check its props, as from memory, it will cause some problems.

Can you move that into its own PR and start by implementing the \__new__ method?

Comment (Sentient07) [inline theano/gpuarray/basic_ops.py]: Yeah, i've done that. But should that be a separate PR ?

**Comment by Sentient07** [comment]
There are few redundant commits. #128f257 and #3855671 (Did in #4570 but with a different name),  	#5bbbb31 did as a different PR. Shall I squash them ?

**Comment by lamblin** [comment]
We will need to the the automatic cache (in Op's metaclass) first, and then go on with the cleaning up.
We will not be able to do that for 0.9, so bumping to 0.10

**Comment by Sentient07** [comment]
Closing as continued in https://github.com/Theano/Theano/pull/5579
